src/Common/GuardedPoolAllocator.cpp

```diff
@@ -0,0 +1,328 @@
+    assert(user_ptr + size <= slot_end);
+
+    /// If a slot is multiple pages in size, and the allocation takes up a single
+    /// page, we can improve overflow detection by leaving the unused pages as
+    /// unmapped. Thus we do set PROT_READ | PROT_WRITE on entire slot. Instead
+    /// we set those only to pages, which will be used
+    const size_t page_size = state.page_size;
+    allocateInGuardedPool(reinterpret_cast<void *>(getPageAddr(user_ptr, page_size)), roundUpTo(size, page_size));
+
+    meta->recordAllocation(user_ptr, size);
+    meta->allocation_trace.recordBacktrace();
+
+    /// I want to remove this output at the last moment before merge
+    // printf(
+    //     "Given out: \n  slot=%zu\n  address=(%p)\n  (start=%p, end=%p)\n",
+    //     slot_index,
+    //     reinterpret_cast<void *>(user_ptr),
+    //     reinterpret_cast<void *>(user_ptr),
+    //     reinterpret_cast<void *>(user_ptr + roundUpTo(size, page_size)));
+    // printf("user_ptr=%p, slot_end=%p\n", reinterpret_cast<void *>(user_ptr), reinterpret_cast<void *>(slot_end));
+    // printf("Distance to guard page: %zu\n", slot_end - user_ptr);
+
+    // printf("\n");
+    return reinterpret_cast<void *>(user_ptr);
+}
+
+void GuardedPoolAllocator::trapOnAddress(uintptr_t address, Error err)
+{
+    state.failure_type = err;
+    state.failure_address = address;
+
+    /// Raise SIGSEGV by touching first guard page.
+    volatile char * p = reinterpret_cast<char *>(state.guarded_page_pool);
+    *p = 0;
+
+    __builtin_trap();
+}
+
+void GuardedPoolAllocator::deallocate(void * ptr)
+{
+    fflush(stdout);
+    uintptr_t u_ptr = reinterpret_cast<uintptr_t>(ptr);
+    size_t slot = state.getNearestSlot(u_ptr);
+    uintptr_t slot_start = state.slotToAddr(slot);
+    AllocationMetadata * meta = addrToMetadata(u_ptr);
+
+    if (meta->addr != u_ptr)
+    {
+        /// If multiple errors occur at the same time, use the first one.
+        std::lock_guard guard{pool_reservation_mutex};
+        trapOnAddress(u_ptr, Error::INVALID_FREE);
+    }
+
+    {
+        std::lock_guard guard{pool_reservation_mutex};
+        if (meta->is_deallocated)
+        {
+            trapOnAddress(u_ptr, Error::DOUBLE_FREE);
+        }
+
+        /// Ensure that the deallocation is recorded before marking the page as
+        /// inaccessible. Otherwise, a racy use-after-free will have inconsistent
+        /// metadata.
+        meta->recordDeallocation();
+
+        /// Ensure that the unwinder is not called if the recursive flag is set,
+        /// otherwise non-reentrant unwinders may deadlock.
+        if (!getThreadLocals()->recursive_guard)
+        {
+            ScopedRecursiveGuard recursive_guard;
+            meta->deallocation_trace.recordBacktrace();
+        }
+    }
+
+    deallocateInGuardedPool(reinterpret_cast<void *>(slot_start), state.maximumAllocationSize());
+
+    /// Acquire mutex to free slot
+    std::lock_guard guard{pool_reservation_mutex};
```

Comment:
Why not put this lock into `freeSlotInGuardedPool`?

Reply:
Ok, done the same with `reserveSlotInGuardedPool`.